docs/enhancements/fep-001-api-fleet-templating.md

```diff
@@ -0,0 +1,257 @@
+## Summary
+
+The Fleet API is designed to manage big groups of devices that share configuration
+and operating system image.
+
+Some use cases point to the necessity of customizing some of those details based
+on the device's metadata.
+
+## Motivation
+
+Allowing templating of the device spec in the fleet API we could achieve a more
+flexible management, allowing customers to manage devices in a more granular way.
+
+### Goals
+
+* Define templating mechanism for the device spec template configuration in fleets,
+  based on labels and other metadata.
+
+* Enable different os images based on the device images.
+```yaml
+  template:
+    spec:
+      os:
+        image: access-control-{{ .device.labels["region"] }}:1.0.0
+```
+
+* Enable different references for a config source based on labels: 
+```yaml
+- name: microshift-manifests
+  configType: GitConfigProviderSpec
+  gitRef:
+    repository: basic-nginx-demo
+    targetRevision: 1.0.0-{{ .device.labels["region"] }}
+    path: /basic-nginx-demo/configuration
+```
+
+* Enable different folders withing a config source based on labels: 
+
+```yaml
+- name: microshift-manifests
+  configType: GitConfigProviderSpec
+  gitRef:
+    repository: basic-nginx-demo
+    targetRevision: 1.0.0
+    path: /basic-nginx-demo/configuration/{{ .device.labels["region"] }}
+```
+
+* Enable different folders withing a config source based on labels: 
+
+```yaml
+- name: microshift-manifests
+  configType: GitConfigProviderSpec
+  gitRef:
+    repository: basic-nginx-demo
+    targetRevision: 1.0.0
+    path: /basic-nginx-demo/configuration/{{ .device.labels["region"] }}/{{ .device.labels["site"] }}
+```
+
+* Enable templating of inline configuration based on device details:
+```yaml
+ - name: motd-update
+   configType: InlineConfigProviderSpec
+   inline:
+     ignition:
+       version: 3.4.0
+     storage:
+       files:
+         - contents:
+             source: >-
+               data:,Device%20{{ .device.name }}%0AThis%20system%20is%20managed%20by%20flightctl.%0A
+           mode: 422
+           overwrite: true
+           path: "/etc/motd"
+```
+### Non-Goals
+
+## Proposal
+
+### Device API
+
+The device API does not change, in the device API we don't allow templating. Any
+device specifications from a fleet should already be rendered into the device
+
+Variable/template expansion is not supported at the device API level, as:
+  * We are operating on a single device, variables aren't really beneficial
+  * Updates to labels will not affect configuration unless the device is part of a fleet.
+ll be considered offline.
+
+### Fleet API
+
+The template.spec field of a fleet allows the use of go-template logic to render the device spec.
+```yaml
+kind: Fleet
+metadata:
+  name: autonomous-forklifts
+spec:
+  template:
+    spec:
+      os:
+        image: forklifts-{{ .device.metadata.label[factory] }}:latest
+      config:
+        - name: factory-wifi-access-credentials
+          configType: K8sSecret
```

Comment:
Since `K8sSecret` are not yet implemented, this example may not be accurate (actually the current API value is `KubernetesSecretProviderSpec`)